### PR TITLE
 randomUsernameAndPassword - first assignment

### PR DESCRIPTION
**Pull request by randomUsernameAndPassword**

### CODE REVIEW

File: randomUsernameAndPassword/assignment1/Q1.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+def is_anagram(str1, str2, case_sensitive):
+    
+    if not case_sensitive:
+        str1.lower()
+        str2.lower()
+    
+    str1 = str1.split(" ")
+    str2 = str2.split(" ")
+    str1 = sorted([''.join(sorted(i)) for i in str1])
```
Comment (jdb) [inline randomUsernameAndPassword/assignment1/Q1.py]: Short and powerful: very cool/

In this code you duplicate the logic to "normalize" a sentence: do you want to create a function that takes the sentence and return a structure that is comparable?

File: randomUsernameAndPassword/assignment1/Q2.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  def __init__(self, values):
+    assert(values)
+    head, tail = values[0], values[1:]
+    self.data = head
+    if tail:
+      self.next = Node(tail)
+    else:
+      self.next = None
+
+  def get_next(self):
```
Comment (jdb) [inline randomUsernameAndPassword/assignment1/Q2.py]: Please use direct access to attribute, this is common in Python
https://google.github.io/styleguide/pyguide.html?showone=Properties#Properties

What advantages do you see with the getters?

File: randomUsernameAndPassword/assignment1/Q2.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    return self.data
+
+def get_kth_last(head, k):
+  if (head == None or k < 0):
+    return None
+  
+  elem = head
+  counter = 0
+  
+  while (elem is not None):
```
Comment (jdb) [inline randomUsernameAndPassword/assignment1/Q2.py]: Do you think you can add a __len__ method to the Node?

File: randomUsernameAndPassword/assignment1/Q2.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+  def get_next(self):
+    return self.next
+
+  def get_data(self):
+    return self.data
+
+def get_kth_last(head, k):
+  if (head == None or k < 0):
+    return None
```
Comment (jdb) [inline randomUsernameAndPassword/assignment1/Q2.py]: returning None is uncommon and is ambiguous as the k-th element could actually be None and you would have no way to distinguish the two.

It is much more common to raise an Exception:

>>> [1,2,3][10]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: list index out of range

Can you raise an IndexError?

File: randomUsernameAndPassword/assignment1/Q2.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    else:
+      self.next = None
+
+  def get_next(self):
+    return self.next
+
+  def get_data(self):
+    return self.data
+
+def get_kth_last(head, k):
```
Comment (jdb) [inline randomUsernameAndPassword/assignment1/Q2.py]: Do you think you can get the bracket notation to work?

list = Node([4, 2, 5, 6, 9, 8])
list[-2] -> 6

Bracket notation for ordered container is more common, it is easier for your users who do not have to learn a new API for a new container.

To make the bracket notation work, you need to implement __getitem__ on the Node class.

**Review by jdb — CHANGES_REQUESTED** [review]

File: randomUsernameAndPassword/assignment1/Q2.py (hunk 5 above)
Comment (randomUsernameAndPassword) [reply randomUsernameAndPassword/assignment1/Q2.py]: Really nice improvement, thank you!

File: randomUsernameAndPassword/assignment1/Q2.py (hunk 2 above)
Comment (randomUsernameAndPassword) [reply randomUsernameAndPassword/assignment1/Q2.py]: Fixed

File: randomUsernameAndPassword/assignment1/Q2.py (hunk 4 above)
Comment (randomUsernameAndPassword) [reply randomUsernameAndPassword/assignment1/Q2.py]: Fixed

File: randomUsernameAndPassword/assignment1/Q2.py (hunk 3 above)
Comment (randomUsernameAndPassword) [reply randomUsernameAndPassword/assignment1/Q2.py]: Fixed

File: randomUsernameAndPassword/assignment1/Q1.py (hunk 1 above)
Comment (randomUsernameAndPassword) [reply randomUsernameAndPassword/assignment1/Q1.py]: Fixed

File: randomUsernameAndPassword/assignment1/Q2.py [hunk 6]
```diff
@@ -1,41 +1,44 @@
...
+            raise IndexError()
+        if (index < 0):
+            index = index + length
+        
+        elem = self
+        for i in range(index):
+            elem = elem.next
+        return elem
+            
+    def len(self):
```
Comment (jdb) [inline randomUsernameAndPassword/assignment1/Q2.py]: You can __len__ to make your class compatible with the builtin len() Python function.

**Review by jdb — APPROVED** [review]

**Pull request merged**